tests/inc/ustream_compliance_e2e.h

```diff
@@ -0,0 +1,155 @@
+#endif
+
+#if(USTREAM_COMPLIANCE_EXPECTED_CONTENT_LENGTH / 20 == 0)
+#error "USTREAM_COMPLIANCE_EXPECTED_CONTENT_LENGTH must be at least 20 uint8_t long"
+#endif
+
+/* split the content in 4 parts. */
+#define USTREAM_COMPLIANCE_LENGTH_1           (USTREAM_COMPLIANCE_EXPECTED_CONTENT_LENGTH >> 2)
+#define USTREAM_COMPLIANCE_LENGTH_2           (USTREAM_COMPLIANCE_LENGTH_1 + USTREAM_COMPLIANCE_LENGTH_1)
+#define USTREAM_COMPLIANCE_LENGTH_3           (USTREAM_COMPLIANCE_LENGTH_2 + USTREAM_COMPLIANCE_LENGTH_1)
+
+/* create local buffer with enough size to handle the full content. */
+#define USTREAM_COMPLIANCE_TEMP_BUFFER_LENGTH (USTREAM_COMPLIANCE_EXPECTED_CONTENT_LENGTH + 2)
+#define TEST_POSITION 10
+#define TEST_SIZE 10
+
+static AZ_USTREAM* compliance_thread_one_ustream;
+static AZ_USTREAM* compliance_thread_two_ustream;
+static const uint8_t* const compliance_expected_content = (const uint8_t*)USTREAM_COMPLIANCE_EXPECTED_CONTENT;
+
+static int compliance_thread_one_func(void* arg)
+{
+    uint8_t buf_result[USTREAM_COMPLIANCE_EXPECTED_CONTENT_LENGTH] = { 0 };
+    size_t size_result;
+    offset_t cur_pos;
+    az_ustream_read(
+        compliance_thread_one_ustream,
+        buf_result,
+        TEST_SIZE,
+        &size_result);
+    az_ustream_get_position(compliance_thread_one_ustream, &cur_pos);
+    ASSERT_ARE_EQUAL(int, TEST_SIZE, cur_pos);
+    ASSERT_ARE_EQUAL(int, 0, strncmp((const char*)USTREAM_COMPLIANCE_LOCAL_EXPECTED_CONTENT, (const char*)buf_result, TEST_SIZE));
+
+    return 0;
+}
+
+static int compliance_thread_two_func(void* arg)
+{
+    uint8_t buf_result[USTREAM_COMPLIANCE_EXPECTED_CONTENT_LENGTH] = { 0 };
+    size_t size_result;
+    offset_t cur_pos;
+    size_t position = USTREAM_COMPLIANCE_LENGTH_1;
+    az_ustream_set_position(compliance_thread_two_ustream, TEST_POSITION);
+    az_ustream_read(
+        compliance_thread_two_ustream,
+        buf_result,
+        TEST_SIZE,
+        &size_result);
+    az_ustream_get_position(compliance_thread_two_ustream, &cur_pos);
+    ASSERT_ARE_EQUAL(int, TEST_POSITION + TEST_SIZE, cur_pos);
+    ASSERT_ARE_EQUAL(int, 0, strncmp((const char*)(USTREAM_COMPLIANCE_LOCAL_EXPECTED_CONTENT + TEST_POSITION), (const char*)buf_result, TEST_SIZE));
+
+    return 0;
+}
+
+/*
+ * Start compliance tests:
+ */
+TEST_FUNCTION(az_ustream_e2e_compliance_multi_read_succeed)
+{
+    ///arrange
+    AZ_USTREAM* multi_ustream = USTREAM_COMPLIANCE_TARGET_FACTORY;
+    ASSERT_IS_NOT_NULL(multi_ustream);
+    AZ_USTREAM* concat_ustream = USTREAM_COMPLIANCE_TARGET_FACTORY;
+    ASSERT_IS_NOT_NULL(concat_ustream);
+
+    AZ_USTREAM_MULTI_DATA_CB* multi_data1 =
+        (AZ_USTREAM_MULTI_DATA_CB*)malloc(sizeof(AZ_USTREAM_MULTI_DATA_CB));
+    ASSERT_IS_NOT_NULL(multi_data1);
+
+    AZ_ULIB_RESULT result = az_ustream_concat(multi_ustream, concat_ustream, multi_data1, free);
+    ASSERT_ARE_EQUAL(int, result, AZ_ULIB_SUCCESS);
+
+    az_ustream_dispose(concat_ustream);
+    free(concat_ustream);
```

Comment:
How do you know that you should free this memory. Consider add another required function to release this memory. Symmetric of USTREAM_COMPLIANCE_TARGET_FACTORY.

Reply:
Interesting. We do not free this memory on the UT.

---
In reply to: [333155887](https://github.com/Azure/azure-ulib-c/pull/34#discussion_r333155887) [](ancestors = 333155887)

Reply:
yea so i try and make it clear on line 37 of `ustream_e2e_sample.c` that in the e2e case the instance will be malloc'd. the other option is to have the factory take an input parameter of a ustream instance. i can make that change for the e2e and ut case since it isn't currently done that way.

---
In reply to: [333157270](https://github.com/Azure/azure-ulib-c/pull/34#discussion_r333157270) [](ancestors = 333157270,333155887)

Reply:
I would add another function USTREAM_COMPLIANCE_DESTROY_TARGET.

---
In reply to: [333158677](https://github.com/Azure/azure-ulib-c/pull/34#discussion_r333158677) [](ancestors = 333158677,333157270,333155887)

Reply:
this has been updated to what we talked about

---
In reply to: [333159996](https://github.com/Azure/azure-ulib-c/pull/34#discussion_r333159996) [](ancestors = 333159996,333158677,333157270,333155887)